test_plans/reference_semantics.asciidoc

```diff
@@ -0,0 +1,92 @@
+:sectnums:
+:xrefstyle: short
+
+= Test plan for common reference semantics
+
+This is a test plan for the common reference semantics as described in Section 4.5.2. of the SYCL 2020 specification. The estimated development time is two days.
+
+== Testing scope
+
+No negative test are included.
+
+=== Backend coverage
+
+All the tests described below are not backend-specific and are performed for any SYCL backend.
+
+=== Device coverage
+
+All tests construct a test device for which conformance is assessed. All the tests described below are performed once for that test device.
+
+== Tests
+
+The common reference semantics are defined for the following runtime classes in the `sycl` namespace (Section 4.5.2.):
+`accessor`, `buffer`, `context`, `device`, `device_image`, `event`, `host_accessor`, `host_sampled_image_accessor`, `host_unsampled_image_accessor`, `kernel`, `kernel_id`, `kernel_bundle`, `local_accessor`, `platform`, `queue`, `sampled_image`, `sampled_image_accessor`, `stream`, `unsampled_image`, and `unsampled_image_accessor`.
+
+Of these, the following classes are legal kernel parameter types (Section 4.12.4.): `accessor` (when templated with `target::device` or the deprecated `target::global_buffer`, `target::constant_buffer`, or `target::local`), `local_accessor`, `unsampled_image_accessor` (when templated with `image_target::device`), `sampled_image_accessor` (when templated with `image_target::device`), `stream`.
+
+
+The below tests are performed on the host for the following types:
+
+- `accessor<int, 1, access_mode::read_write, target::host_task>`
+- `buffer<int>`
+- `context`
+- `device`
+- `device_image<bundle_state::executable>`
+- `event`
+- `host_accessor<int>`
+- `host_sampled_image_accessor<int, 1>`
+- `host_unsampled_image_accessor<int>`
+- `kernel`
+- `kernel_id`
+- `kernel_bundle<bundle_state::executable>`
+- `platform`
+- `queue`
+- `sampled_image`
+- `unsampled_image`
+
+The below tests are performed in a kernel for the following types:
+
+- `accessor<int, 1, access_mode::read_write, target::device>`
+- `local_accessor<int, 0>`
+- `unsampled_image_accessor<int, 1, access_mode::read>`
+- `sampled_image_accessor<int, 1>`
+- `stream`
```

Comment:
All of these types can be used to construct an object on the host too.

Reply:
I have re-written the tests: there are "host types" and "kernel types". The tests for the equality operators and `std::hash` are done in a host application for host and kernel types. The tests for type traits are done in a host application for host and kernel types and in a kernel function for kernel types only.

I do not think it is useful to call the copy/move constructor/assignment functions from a kernel function, since the return value cannot be checked as equality is not guaranteed to be available.